pkg/controlplane/controller.go

```diff
@@ -88,10 +89,19 @@ type Controller struct {
 }
 
 // NewBootstrapController returns a controller for watching the core capabilities of the master
-func (c *completedConfig) NewBootstrapController(legacyRESTStorage corerest.LegacyRESTStorage, serviceClient corev1client.ServicesGetter, nsClient corev1client.NamespacesGetter, eventClient corev1client.EventsGetter, readyzClient rest.Interface) *Controller {
+func (c *completedConfig) NewBootstrapController(legacyRESTStorage corerest.LegacyRESTStorage, serviceClient corev1client.ServicesGetter, nsClient corev1client.NamespacesGetter, eventClient corev1client.EventsGetter, readyzClient rest.Interface) (*Controller, error) {
 	_, publicServicePort, err := c.GenericConfig.SecureServing.HostPort()
 	if err != nil {
-		klog.Fatalf("failed to get listener address: %v", err)
+		return nil, fmt.Errorf("failed to get listener address: %w", err)
+	}
+
+	// if the bootstrap controller reconcile the kubernetes.default Service and Endpoints
+	// it must guarantee that both have the same IP family or it will not work because the
+	// Service is single-stack
+	if c.ExtraConfig.EndpointReconcilerType != reconcilers.NoneEndpointReconcilerType {
+		if netutils.IsIPv4CIDR(&c.ExtraConfig.ServiceIPRange) != netutils.IsIPv4(c.GenericConfig.PublicAddress) {
```

Comment:
Is it required that the apiserver Service use `ServiceIPRange` and not `SecondaryServiceIPRange` ?  Should we check `ExtraConfig.APIServerServiceIP` rather than the range?

Reply:
And if not, let's comment here why...

Reply:
right now is single stack, @danwinship will make it dual https://github.com/kubernetes/enhancements/issues/2438 but it will take some time.

I can add a better comment though